internal/service/service.go

```diff
@@ -0,0 +1,271 @@
+package service
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"path/filepath"
+	"reflect"
+	"sync"
+
+	"github.com/google/uuid"
+	"github.com/gopheryan/jobby/job"
+	"github.com/gopheryan/jobby/jobmanagerpb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const defaultOutputBufferSize = 4096
+
+type UserGetter interface {
+	GetUserContext(context.Context) string
+}
+
+type JobIdGetter interface {
+	GetJobId() []byte
+}
+
+type jobData struct {
+	// User who owns the job
+	Owner string
+	Job   *job.Job
+}
+
+type Jobby struct {
+	jobmanagerpb.UnimplementedJobManagerServer
+	// minimum buffer size use when reading from job output
+	// Must be > 0. Internal reader will block until 'minimumRead'
+	// bytes of data have been read from the job before sending
+	// output to the caller
+	minimumRead int
+	// Used to determine which user a request is coming from
+	// decouples the service from the auth strategy
+	userGetter UserGetter
+	// Base directory in which to store output files for jobs
+	directory string
+	// Keep track of jobs!
+	// used as: map[uuid.UUID]*jobData
+	jobDirectory sync.Map
+}
+
+func NewJobService(minRead int, userGetter UserGetter, dir string) *Jobby {
+	return &Jobby{
+		minimumRead: minRead,
+		userGetter:  userGetter,
+		directory:   dir,
+	}
+}
+
+func (j *Jobby) Register(srv *grpc.Server) {
+	srv.RegisterService(&jobmanagerpb.JobManager_ServiceDesc, j)
+}
+
+func (j *Jobby) GetJobOutput(req *jobmanagerpb.GetJobOutputRequest, srv jobmanagerpb.JobManager_GetJobOutputServer) error {
+	subLogger := slog.With("user", j.userGetter.GetUserContext(srv.Context()), "request", req)
+	subLogger.Info("Handling 'GetJobOutput' request")
+
+	jobData, st := j.getJob(srv.Context(), req)
+	if st != nil {
+		return st.Err()
+	}
+
+	var reader io.ReadCloser
+	var err error
+	if req.Type == jobmanagerpb.OutputType_OUTPUT_TYPE_STDOUT {
+		reader, err = jobData.Job.Stdout()
+	} else if req.Type == jobmanagerpb.OutputType_OUTPUT_TYPE_STDERR {
+		reader, err = jobData.Job.Stderr()
+	} else {
+		return status.Error(codes.InvalidArgument, "Must specify valid output type")
+	}
+	if err != nil {
+		return status.Error(codes.Internal, "Error attaching to job output")
+	}
+
+	// The caller can cancel/detach at any time. This cancellation is communicated
+	// to this handler via context cancellation
+	context.AfterFunc(srv.Context(), func() {
+		subLogger.Info("GetJobOutput request cancelled")
+		// One call to 'Close' will shut down this whole operation...
+		// This is going to cause an error on our reader
+		if err = reader.Close(); err != nil {
+			subLogger.Error("Error closing job output reader", slog.String("error", err.Error()))
+		}
+
+	})
```

Comment:
Should the stop function returned from AfterFunc be called prior to exiting the function?
```suggestion
	stop := context.AfterFunc(srv.Context(), func() {
		subLogger.Info("GetJobOutput request cancelled")
		// One call to 'Close' will shut down this whole operation...
		// This is going to cause an error on our reader
		if err = reader.Close(); err != nil {
			subLogger.Error("Error closing job output reader", slog.String("error", err.Error()))
		}
	})
	defer stop()
```

Reply:
I can't argue against tidying up.